elastic-buffer-sim/exe/Main.hs

```diff
@@ -2,70 +2,644 @@
+      Random n      -> (,dName) <$> (randomGraph n >>= plotGraph settings)
+      DotFile f     -> (fromDot <$> readFile f) >>= \case
+        Right (g, name) -> (,name) <$> plotGraph settings g
+        Left err        -> die $ "ERROR: Invalid DOT file - " <> f <> "\n" <> err
+    Nothing ->
+      handleParseResult $ Failure
+        $ parserFailure defaultPrefs cliParser (ShowHelpText Nothing) []
+
+  let topologyFile = outDir </> "topology.gv"
+  writeFile topologyFile $ (<> "\n") $ render $ toDot g name
+  BS.writeFile (outDir </> "simulate.json") $ encode options
+     { stable   = isStable
+     , offsets  = offs
+     , jsonArgs = Nothing
+     , topology = case topology of
+         Just (Random _) -> Just $ DotFile topologyFile
+         _               -> topology
+     }
+
+  if isStable
+  then exitSuccess
+  else exitFailure
+
+randomGraph :: Int -> IO Graph
+randomGraph n = do
+  let is = [0,1..n - 1]
+
+  -- get some random vertex permuation for ensuring connectivity
+  aP <- newListArray (0, n - 1) is
+  replicateM_ (10 * n) $ do
+    p1 <- randomRIO (0, n - 1)
+    p2 <- randomRIO (0, n - 1)
+    v1 <- readArray aP p1
+    v2 <- readArray aP p2
+    writeArray aP p1 v2
+    writeArray aP p2 v1
+
+  randomPermutation <- getElems (aP :: IOUArray Int Int)
+
+  -- get some random edge availability matrix for self-loop-free,
+  -- undirected graphs
+  aE <- replicateM (n * n) randomIO >>= newListArray ((0, 0), (n - 1, n - 1))
+  forM_ is $ \i ->
+    writeArray aE (i, i) False
+  forM_ [(i, j) | i <- is, j <- is, i /= j] $ \(i, j) -> do
+    d1 <- readArray aE (i, j)
+    d2 <- readArray aE (j, i)
+    when (d1 /= d2) $ do
+      writeArray aE (i, j) False
+      writeArray aE (j, i) False
+
+  -- ensure connectivity
+  forM_ (unchain randomPermutation) $ \(i, j) -> do
+    writeArray aE (i, j) True
+    writeArray aE (j, i) True
+
+  available <- freeze (aE :: IOUArray (Int, Int) Bool)
+
+  -- create the graph
+  return $ buildG (0, n - 1)
+    [ (i, j)
+    | i <- is
+    , j <- is
+    , available A.! (i, j)
+    ]
+
+toDot :: Graph -> String -> (Bool, GraphType, Maybe Name, [Statement])
+toDot g name = (True, Graph, Just $ StringID name, map asEdgeStatement $ edges g)
+ where
+  asEdgeStatement (x, y) =
+    EdgeStatement (map (\i -> NodeRef (XMLID ('n' : show i)) Nothing) [x,y]) []
+
+fromDot :: String -> Either String (Graph, String)
+fromDot cnt = do
+  (strict, gType, maybe "" fromDotName -> name, statements) <- parse cnt
+
+  when (not strict) $ Left "Graph must be strict"
+  when (gType == Digraph) $ Left "Graph must be undirected"
+
+  namedEdges <-
+    fmap
+      ( concatMap (map (\(x,y) -> (asString x, asString y)) . unchain)
+      . catMaybes
+      )
+    $ forM statements $ \case
+        EdgeStatement xs _ -> fmap Just $ forM xs $ \case
+          NodeRef n _ -> return n
+          Subgraph{}  -> Left "Subgraphs are not supported"
+        _ -> pure Nothing
```

Comment:
Could you pull apart these two phases into two variables and document what's happening? I can't quite follow all the monads/forM without concentrating. Seems like:

1. You're extracting all edges
2. Edges can consist of multiple nodes in dot, so you're decomposing these edges into "2-node" edges.

So I think I'd split this up like:

```haskell
-- Extract all edges from dot file. Note that dot edges can consist of more
-- than 2 nodes. We later flatten this to 2-node edges.
mutltiEdges <-
  fmap catMaybes $
    forM statements $ \case
      ..

let 
  edges = concatMap pairwise multiEdges
  namedEdges = map (bimap asString asString) twoEdges
```

The intermediate variable names should help the reader understand your code. (Not saying I picked the best names, or that this code even compiles.)

Reply:
Good point. I'll update it.